libpromises/rlist.c

```diff
@@ -325,18 +325,56 @@ Rlist *RlistAppendRval(Rlist **start, Rval rval)
     return rp;
 }
 
-Rval RvalNew(const void *item, RvalType type)
+Rval RvalNewRewriter(const void *item, RvalType type, JsonElement *map)
 {
     switch (type)
     {
     case RVAL_TYPE_SCALAR:
-        return (Rval) { xstrdup(item), RVAL_TYPE_SCALAR };
+        if (NULL != map && JsonLength(map) > 0)
+        {
+            // TODO: replace with BufferSearchAndReplace when the
+            // string_replace code is merged.
+            // Sorry about the CF_BUFSIZE ugliness.
+            int max_size = 10*CF_BUFSIZE+1;
+            const char* buffer_from = malloc(max_size);
+            const char* buffer_to = malloc(max_size);
+
+            Buffer *format = BufferNew();
+            strncpy(buffer_from, item, max_size);
+
+            JsonIterator iter = JsonIteratorInit(map);
+            const char *key;
+
+            for (int loop=0; loop < 3; loop++)
+            {
+                while ((key = JsonIteratorNextKey(&iter)))
+                {
+                    BufferPrintf(format, "$(%s)", key);
+                    ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+                    strncpy(buffer_from, buffer_to, max_size);
+                    BufferPrintf(format, "${%s}", key);
+                    ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+                }
+            }
```

Comment:
I've modified the code to break early and to do 10 iterations, but would prefer if you make the optimizations you propose after it's merged.